internal/slicesx/heap.go

```diff
@@ -0,0 +1,119 @@
+// Copyright 2020-2024 Buf Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package slicesx
+
+import "cmp"
+
+// Heap is a port of Go's [container/heap] package to use generics instead of
+// interface calls. Instead, entries consist of an ordered key and an arbitrary
+// value.ß
+//
+// A zero heap is empty and ready to use.
+type Heap[K cmp.Ordered, V any] struct {
+	keys []K
+	vals []V
+}
+
+// Len returns the number of elements in the heap.
+func (h *Heap[K, V]) Len() int {
+	return len(h.keys)
+}
+
+// Push pushes the element x onto the heap.
+func (h *Heap[K, V]) Push(k K, v V) {
+	h.push(k, v)
+	h.up(h.Len() - 1)
+}
+
+// Pop removes and returns the entry with the least key from the heap.
+//
+// Pop is equivalent to [Heap.Remove](0).
+func (h *Heap[K, V]) Pop() (K, V) {
+	n := h.Len() - 1
+	h.swap(0, n)
+	h.down(0, n)
+	return h.pop()
+}
+
+// Remove removes and returns the entry at index i from the heap.
+func (h *Heap[K, V]) Remove(i int) (K, V) {
+	n := h.Len() - 1
+	if n != i {
+		h.swap(i, n)
+		if !h.down(i, n) {
+			h.up(i)
+		}
+	}
+	return h.pop()
+}
+
+// Fix re-establishes the heap invariant after the key of the nth entry changes.
+func (h *Heap[K, V]) Fix(n int) {
+	if !h.down(n, h.Len()) {
+		h.up(n)
+	}
+}
+
+func (h *Heap[K, V]) up(j int) {
+	for {
+		i := (j - 1) / 2 // parent
+		if i == j || !h.less(j, i) {
+			break
+		}
+		h.swap(i, j)
+		j = i
+	}
+}
+
+func (h *Heap[K, V]) down(i0, n int) bool {
+	i := i0
+	for {
+		j1 := 2*i + 1
+		if j1 >= n || j1 < 0 { // j1 < 0 after int overflow
+			break
+		}
+		j := j1 // left child
+		if j2 := j1 + 1; j2 < n && h.less(j2, j1) {
+			j = j2 // = 2*i + 2  // right child
+		}
+		if !h.less(j, i) {
+			break
+		}
+		h.swap(i, j)
+		i = j
+	}
+	return i > i0
+}
+
+func (h *Heap[K, V]) less(i, j int) bool {
+	return cmp.Compare(h.keys[i], h.keys[j]) < 0
```

Comment:
I don't think the function call is necessary. Thanks to the `cmp.Ordered` constraint, you can just do `h.keys[i] < h.keys[j]`.

I think `cmp.Compare` is mainly provided as a comparator -- so if the data structure supported `K any` with a custom comparator function, callers could supply `cmp.Compare` if the key is ordered. But since we require the key to be ordered, we don't need that level of ceremony.

Reply:
Looks like `cmp.Compare` also handles NaN for float types, to provide a stable ordering where NaN values are less than all others. But do we really care about that? I appreciate the intent to make this a very general-purpose container type, but I think we're better off tailoring it to what we need and omitting what we don't.